### PR TITLE
Passes full file path to the replacement function

### PR DESCRIPTION
**Pull request by asins**
#67 finishing version

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "gulp-replace",
-  "version": "0.5.4",
+  "version": "0.5.5",
```
Comment (lazd) [inline package.json]: This is a new feature, so we should bump to `0.6.0`.

**Comment by asins** [comment]
Resolved after performing `gulp.dest()` method to modify `this` pointer is a problem.

**Comment by asins** [comment]
Resolved after performing `gulp.dest()` method to modify `this` pointer is a problem.

**Comment by lazd** [comment]
@asins I'm not sure what you mean by:

> `gulp.dest()` method to modify this pointer is a problem.

Can you elaborate?

**Comment by asins** [comment]
demo:

```
npm install gulp gulp-replace --save-dev
```

### add code for `gulp-replace/index.js`: 
```
        if (options.passFileName && typeof replacement === 'function') {
          var userReplacement = replacement;
          replacement = function () {
			  var context = {
				  filePath: file.path
			  };
			  return userReplacement.apply(context, arguments);
		  };
        }
```

### in demo folid `tree -L 1`

```
├── gulpfile.js
├── node_modules
├── package.json
└── src
```

### The `gulpfile.js`:

```js
const Fs = require('fs');
const Path = require('path');
const gulp = require('gulp');
const replace = require('gulp-replace');

var i = 0;
gulp.task('tests', function () {
	return gulp.src('./src/css/*.css')
		// .pipe(less())
		.pipe(replace(/url\((["']?)([^\1\)]+)\1\)/g, cssAssets, {passFileName: true}))
		.pipe(gulp.dest('./build/css'))
		// .pipe(cssnano())
		.pipe(gulp.dest('./dist/css'))

	function cssAssets($0, $1, url){
		console.log('------------>', i++, this.filePath, url);
		return $0;
	}
});
```

### The css files:

```css
/* src/css/a.css */
@font-face {font-family: "iconfont";
  src: url('../fonts/iconfont.eot');
  src: url('../fonts/iconfont.eot?#iefix') format('embedded-opentype'),
  url('../fonts/iconfont.woff') format('woff'),
  url('../fonts/iconfont.ttf') format('truetype'),
  url('../fonts/iconfont.svg#iconfont') format('svg');
}
.YT-has-passport{
	background: url("../images/alipayToYT.jpg") no-repeat center bottom #0496d1;
	background-size:100%;
}

/*src/css/b.css*/
@font-face {font-family: "iconfonts";
	src: url('../fonts/iconfont.eot');
	src: url('../fonts/iconfont.eot?#iefix') format('embedded-opentype'),
	url('../fonts/iconfont.woff') format('woff'),
	url('../fonts/iconfont.ttf') format('truetype'),
	url('../fonts/iconfont.svg#iconfont') format('svg');
}
@font-face{font-family: "asins-font";
	src: url('../fonts/asins.eot');
	src: url('../fonts/asins.eot?#iefix') format('embedded-opentype'),
	url('../fonts/asins.woff') format('woff'),
	url('../fonts/asins.ttf') format('truetype'),
	url('../fonts/asins.svg#iconfont') format('svg');
}
.icon-logo{margin: 40px auto; background:#e6e6e6 url(../images/icon-risk.png) 50% 50% no-repeat;}
.nc-container{ background: url(../images/loading.gif) 50% 50% no-repeat; height: 40px;}
```

### run

Now, run `gulp tests`, you can see after `gulp.dest()` output after the first file `this.filePath` was changed.

```
~/gulptest » gulp tests
[09:40:49] Using gulpfile ~/gulptest/gulpfile.js
[09:40:49] Starting 'tests'...
------------> 0 /Users/asins/gulptest/src/css/a.css ../fonts/iconfont.eot
------------> 1 /Users/asins/gulptest/src/css/a.css ../fonts/iconfont.eot?#iefix
------------> 2 /Users/asins/gulptest/src/css/a.css ../fonts/iconfont.woff
------------> 3 /Users/asins/gulptest/src/css/a.css ../fonts/iconfont.ttf
------------> 4 /Users/asins/gulptest/src/css/a.css ../fonts/iconfont.svg#iconfont
------------> 5 /Users/asins/gulptest/src/css/a.css ../images/alipayToYT.jpg
------------> 6 /Users/asins/gulptest/build/css/a.css ../fonts/iconfont.eot
------------> 7 /Users/asins/gulptest/build/css/a.css ../fonts/iconfont.eot?#iefix
------------> 8 /Users/asins/gulptest/build/css/a.css ../fonts/iconfont.woff
------------> 9 /Users/asins/gulptest/build/css/a.css ../fonts/iconfont.ttf
------------> 10 /Users/asins/gulptest/build/css/a.css ../fonts/iconfont.svg#iconfont
------------> 11 /Users/asins/gulptest/build/css/a.css ../fonts/alinc.eot
------------> 12 /Users/asins/gulptest/build/css/a.css ../fonts/alinc.eot?#iefix
------------> 13 /Users/asins/gulptest/build/css/a.css ../fonts/alinc.woff
------------> 14 /Users/asins/gulptest/build/css/a.css ../fonts/alinc.ttf
------------> 15 /Users/asins/gulptest/build/css/a.css ../fonts/alinc.svg#iconfont
------------> 16 /Users/asins/gulptest/build/css/a.css ../images/icon-risk.png
------------> 17 /Users/asins/gulptest/build/css/a.css ../images/loading.gif
[09:40:49] Finished 'tests' after 31 ms
```
[demo.zip](https://github.com/lazd/gulp-replace/files/535106/demo.zip)

**Comment by asins** [comment]
I found that filepath was the first time the file was sent to multiple files, because the function was rewritten several times, and the internal value was not changed.

**Comment by lazd** [comment]
A couple notes: 

1. We should be consistent in referring to this as the `filePath` as is contains all of the path information, not just the name. Vinyl uses `file.path`, so perhaps it should be `this.path`.

2. Perhaps we should just make the Vinyl file object the context with `replacement = userReplacement.bind(file);`. Then all of the instance [methods](https://github.com/gulpjs/vinyl#instance-methods) and [properties](https://github.com/gulpjs/vinyl#instance-properties) of the Vinyl file object will be available, not just the path. That would also result in 1. above being satisfied.

Thoughts?

**Comment by lazd** [comment]
> I found that filepath was the first time the file was sent to multiple files, because the function was rewritten several times, and the internal value was not changed.

Please provide tests with a single file and multiple files. If you wrote the tests before the code\, you wouldn't have had this bug 😛  🤓

**Comment by lazd** [comment]
I went ahead and used the approach in #80 for this. Thanks for your contribution; though it wasn't merged, it still helped define the API and the project is better for it.

**Pull request closed without merging**